numba/np/arrayobj.py

```diff
@@ -1562,6 +1562,35 @@ def array_T(context, builder, typ, value):
     return impl_ret_borrowed(context, builder, typ, res)
 
 
+@overload(np.rot90)
+def numpy_rot90(arr, k=1):
+    # supporting axes argument it needs to be included in np.flip
+    if not isinstance(k, (int, types.Integer)):
+        raise errors.TypingError('The second argument "k" must be an integer')
+    if not isinstance(arr, types.Array):
+        raise errors.TypingError('The first argument "arr" must be an array')
+
+    if arr.ndim < 2:
+        raise ValueError('Input must be >= 2-d.')
+
+    axes_list = (1, 0, *range(2, arr.ndim))
+
+    def impl(arr, k=1):
+        k = k % 4
+        if k == 0:
+            return arr[:]
+        elif k == 1:
+            return np.transpose(np.fliplr(arr), axes_list)
+        elif k == 2:
+            return np.flipud(np.fliplr(arr))
+        elif k == 3:
+            return np.fliplr(np.transpose(arr, axes_list))
+        else:
+            assert 0  # unreachable
```

Comment:
i thk this should explicitly raise to prevent `python -O` causing the `assert` to be removed and falling through as a return None, which will alter the return type of the function.

Reply:
The `else` case is entered only when `k = k %  4` is not in `[0, 1, 2, 3]`, so when modulo operation fails. So an alternative is to change the last `elif k==3` to an `else` and  removing lines 1588-1589. In case you prefer to keep the unreachable else case, could you please help me which kind of error I should use?

Reply:
Putting a `raise AssertionError` is suffice. 

Details:

Since this code is compiled by numba, static analysis does not know the range of possible `k`. Type inference (a static analysis) will consider all branches in determining the return type. With `python -O`, the `assert` statement will be omitted; thus, the `else` branch becomes equivalent to a `return None`. The type-inference will then think the output type can be `None` and using an `Optional[Array]` return type unnecessarily.

Reply:
I've just fixed it and thx for the helpful explanation.

Reply:
Thanks @sklam I need to remember to stop doing this C-like idiom :)